### ISSUE TITLE
getpid returns fault pid everytime in some code

### ISSUE THREAD

**Issue by zylthinking** [golang/go#30858]
### What version of Go are you using (`go version`)?

<pre>
1.12

</pre>

### Does this issue reproduce with the latest release?
YES

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/zylthinking/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/zylthinking/go"
GOPROXY=""
GORACE=""
GOROOT="/opt/go"
GOTMPDIR=""
GOTOOLDIR="/opt/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build351775018=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

use cgo to call a c function
```C
/*
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <assert.h>
#include <sys/wait.h>

static void move_cgroup_to(char* pathname, pid_t pid)
{
    char buf[1024];
    int n = strlen(pathname) - 1;
    for (; pathname[n] == '\n'; --n) {
        pathname[n] = 0;
    }
    sprintf(buf, "%s/cgroup.procs", pathname);

    FILE* file = fopen(buf, "a");
    if (file == NULL) {
        return;
    }
    fprintf(file, "%d\n", pid);
    fclose(file);
}

int spawn_emulator_daemon(char* cgroup, char* core, char* mem, char* emu, char* name)
{
    pid_t pid = vfork();
    if ((pid_t) -1 == pid) {
        return -1;
    }

    if (pid > 0) {
        free((void *) cgroup);
        free((void *) core);
        free((void *) mem);
        free((void *) emu);
        free((void *) name);
        do {
            int n = waitpid(pid, NULL, 0);
            if (n != -1 || EINTR != errno) {
                break;
            }
        } while (1);
        return pid;
    }

    if ((pid_t) -1 == pid) {
        _exit(-1);
    }

    pid = vfork();
    if ((pid_t) -1 == pid) {
       _exit(-1);
    }

    if (pid > 0) {
       _exit(0);
    }

    // 虽然在第二次 vfork之前调用 setsid,
    // 在这里调用 getpid 更合适
    // 但是因为 cgo 有bug, 若在这里调用 getpid
    // 返回的是居然是爷爷进程的 pid
    // 做试验直接在 go main函数里面直接调用这个函数却没问题
    // 所以就不知道出这个问题是受哪里的影响
    // 但 setsid 返回的 pid 却是准的
    // 因此将 setsid 直接挪到这里来, 避免调用 getpid
    pid = setsid();
    assert(pid != -1);

    printf("%d %d\n", pid, getpid());
    if (*cgroup != 0 && *cgroup != '\n') {
        move_cgroup_to(cgroup, pid);
    }

    char* argv[] = {
        emu,
        "-cores",
        core,
        "-memory",
        mem,
        "-avd",
        name,
        "-no-boot-anim",
        "-camera-back",
        "none",
        "-camera-front",
        "none",
        "-no-window",
        "-no-sim",
        "-netfast",
        "-noaudio",
        "-partition-size",
        "2048",
        NULL
    };
    execv(emu, (char* const*) argv);
    _exit(-1);
}
*/
```

### What did you expect to see?

getpid returns a pid of itself, same with setsid 

### What did you see instead?

getpid returns grantfather's pid

code above explains in detail with chinese, and it is not occures in a simple 
func main() {
    C.spawn_emulator_daemon(C.CString("aaa"), C.CString("1"), C.CString("2"), C.CString("3"), C.CString("4"));
}

But, the bug  exists abosolutely.
an execution get the output:

[root@docker50 emulator]# ./test -docker
6178 6076

by this line: printf("%d %d\n", pid, getpid());

**Comment by ianlancetaylor** [comment]
As far as I can tell you are asking about the behavior of your C code.  This issue tracker is for Go code, not C code.

That said, your C code calls `vfork` twice in a row.  I would be very surprised if that works.  I suggest that you call `fork` rather than `vfork`.  Read the man page for `vfork`.

**Comment by zylthinking** [comment]
If you check it carefully,  you will find the C code has no problems. It is occures in C code, but the deep reason should be something wrong with CGO, I am not sure here is a proper place for CGO or not.

**Comment by zylthinking** [comment]
And I cant call fork instead of vfork in this case, because I depend the grandfather wake after the father exits, and the father should exit only when the son calls execv.

**Comment by zylthinking** [comment]
OK, I find more information, the following code

    int pid3 = (int) getpid();
    pid = setsid();
    assert(pid != -1);
    int pid2 =  (int) syscall(SYS_getpid);
    printf("%d %d(%d) %d\n", pid, pid3, (int) getpid(), pid2);

 runs in my ubuntu:
10338 10338(10338) 10338

runs in docker(centos7.2) and host centos7.2:
3971 3873(3873) 3971

So, it is seems a docker or OS problem?

**Comment by zylthinking** [comment]
Hello, it maybe the CGO's problem.
I compile the code in go and with c and runs in centos72, and ubuntu

only go version in centos72 get the fault
I have attached the c.go and c.c.
[code.zip](https://github.com/golang/go/files/2970454/code.zip)

**Comment by ianlancetaylor** [comment]
If the difference only happens when calling the C code from Go, then the difference is likely due to the fact that Go programs are multi-threaded.  I don't know what `vfork` does in a multi-threaded program.  The man page says that it suspends the calling thread; I don't know what happens to the other threads.

I don't understand your explanation of why you can't use `fork` instead of `vfork`.
